pkg/cli/cmd/credential/register/azure/sp/serviceprincipal.go

```diff
@@ -0,0 +1,191 @@
+
+import (
+	"context"
+	"fmt"
+
+	v1 "github.com/radius-project/radius/pkg/armrpc/api/v1"
+	"github.com/radius-project/radius/pkg/cli"
+	"github.com/radius-project/radius/pkg/cli/clierrors"
+	"github.com/radius-project/radius/pkg/cli/cmd/commonflags"
+	"github.com/radius-project/radius/pkg/cli/cmd/credential/common"
+	"github.com/radius-project/radius/pkg/cli/connections"
+	cli_credential "github.com/radius-project/radius/pkg/cli/credential"
+	"github.com/radius-project/radius/pkg/cli/framework"
+	"github.com/radius-project/radius/pkg/cli/output"
+	"github.com/radius-project/radius/pkg/cli/workspaces"
+	"github.com/radius-project/radius/pkg/to"
+	ucp "github.com/radius-project/radius/pkg/ucp/api/v20231001preview"
+
+	"github.com/spf13/cobra"
+)
+
+// NewCommand creates an instance of the command and runner for the `rad credential register azure sp` command.
+//
+
+// NewCommand creates a new cobra command for registering an Azure cloud provider service principal credential for a Radius installation,
+// which requires a service principal with the Contributor or Owner role assigned to the provided resource group.
+func NewCommand(factory framework.Factory) (*cobra.Command, framework.Runner) {
+	runner := NewRunner(factory)
+
+	cmd := &cobra.Command{
+		Use:   "sp",
+		Short: "Register (Add or update) Azure cloud provider service principal credential for a Radius installation.",
+		Long: `Register (Add or update) Azure cloud provider service principal credential for a Radius installation.
+
+This command is intended for scripting or advanced use-cases. See 'rad init' for a user-friendly way
+to configure these settings.
+
+Radius will use the provided service principal for all interactions with Azure, including Bicep deployment, 
+Radius Environments, and Radius portable resources. 
+
+Radius will use the provided subscription and resource group as the default target scope for Bicep deployment.
+The provided service principal must have the Contributor or Owner role assigned for the provided resource group
+in order to create or manage resources contained in the group. The resource group should be created before
+calling 'rad credential register azure sp'.
+` + common.LongDescriptionBlurb,
+		Example: `
+# Register (Add or update) cloud provider credential for Azure with service principal authentication
+rad credential register azure sp --client-id <client id/app id> --client-secret <client secret/password> --tenant-id <tenant id>
+`,
+		Args: cobra.ExactArgs(0),
+		RunE: framework.RunCommand(runner),
+	}
+
+	commonflags.AddOutputFlag(cmd)
+	commonflags.AddWorkspaceFlag(cmd)
+
+	cmd.Flags().String("client-id", "", "The client id or app id of an Azure service principal.")
+	_ = cmd.MarkFlagRequired("client-id")
+
+	cmd.Flags().String("client-secret", "", "The client secret or password of an Azure service principal.")
+	_ = cmd.MarkFlagRequired("client-secret")
+
+	cmd.Flags().String("tenant-id", "", "The tenant id of an Azure service principal.")
```

Comment:
Minor suggestion - use `cmd.Flags().StringVar(&r.TenantID, ...)`. Then you don't need to call `cmd.Flags().GetString("tenant-id")`. 

This is a simpler way to write the same code. Most of our CLI code was written before we knew about this feature.

Reply:
ooh snazzy. updated